crates/fuel-gas-price-algorithm/src/v1.rs

```diff
@@ -515,60 +526,69 @@ impl AlgorithmUpdaterV1 {
 
     fn da_block_update(
         &mut self,
-        height_range: Range<u32>,
-        range_cost: u128,
+        heights: &[u32],
+        recording_cost: u128,
     ) -> Result<(), Error> {
-        let expected = self.da_recorded_block_height.saturating_add(1);
-        let first = height_range.start;
-        if first != expected {
-            Err(Error::SkippedDABlock {
-                expected,
-                got: first,
-            })
-        } else {
-            let last = height_range.end.saturating_sub(1);
-            let range_bytes = self.drain_l2_block_bytes_for_range(height_range)?;
-            let new_cost_per_byte: u128 = range_cost.checked_div(range_bytes).ok_or(
-                Error::CouldNotCalculateCostPerByte {
-                    bytes: range_bytes,
-                    cost: range_cost,
-                },
-            )?;
-            self.da_recorded_block_height = last;
-            let new_da_block_cost = self
-                .latest_known_total_da_cost_excess
-                .saturating_add(range_cost);
-            self.latest_known_total_da_cost_excess = new_da_block_cost;
-            self.latest_da_cost_per_byte = new_cost_per_byte;
-            Ok(())
-        }
+        let bytes_for_heights = self.drain_l2_block_bytes_for_heights(heights);
+        match bytes_for_heights {
+            Complete(bytes) => {
+                let new_cost_per_byte: u128 = recording_cost.checked_div(bytes).ok_or(
+                    Error::CouldNotCalculateCostPerByte {
+                        bytes,
+                        cost: recording_cost,
+                    },
+                )?;
+                let new_da_block_cost = self
+                    .latest_known_total_da_cost_excess
+                    .saturating_add(recording_cost);
+                self.latest_known_total_da_cost_excess = new_da_block_cost;
+                self.latest_da_cost_per_byte = new_cost_per_byte;
+            }
+            // If we weren't able to get the bytes for all the heights, we can't update the cost per byte in a logical way.
+            // We will just accept the predicted cost as the real cost and move on with the algorithm
+            // We can safely say that the cost didn't exceed the `recording_cost`, so we will take the
+            // minimum of the two costs
+            Incomplete(bytes) => {
+                let new_guess_recording_cost = bytes
+                    .saturating_mul(self.latest_da_cost_per_byte)
+                    .min(recording_cost);
+                let new_da_block_cost = self
+                    .latest_known_total_da_cost_excess
+                    .saturating_add(new_guess_recording_cost);
+                self.latest_known_total_da_cost_excess = new_da_block_cost;
+            }
```

Comment:
It is not clear to me why missed bundles don't allow us to calculate the cost per byte. We know the total number of bytes and the cost of posting them. Why we can't use these values? What are we afraid of?

Reply:
> missed bundles

No. This is about missed blocks.

> We know the total number of bytes

Not really. We have the total bytes after compression. 

We could in theory go through each block in the bundle and query our api to find the original bytes for each block that's not listed in `unrecorded_blocks`, but this is an edge case that we probably will never hit and I think this solution works well enough since gas price is off-chain and a little error is fine.

Reply:
Question to you: What is `new_cost_per_byte`? Is it the cost per 1 compressed byte submitted by the block committer to DA? Or is it the cost per 1 uncompressed byte for the Fuel block with the compression ratio included?

I would expect the first one to be used. But it seems you use the second in the code.

An example:

Block X
The Fuel Block is - 256KB.
The compressed Fuel Block is - 64KB.
Let's assume the bundle has 1 Fuel block(the block above).
The cost of posting this bundle is 1ETH.

I will expect that `new_cost_per_byte` is `1ETH / 64KB` because it is the actual cost to post bytes to the DA.
The current implementation will use `1ETH / 256KB`. So the `new_cost_per_byte` will be lower - cheaper.

Let's see what the cost will be in the next block, assuming we updated `new_cost_per_byte` by the rules above.

Block X + 1
The Fuel Block is - 256KB.
The projected cost with the current implementation is `256KB * 1 ETH / 256KB = 1 ETH`.
The projected cost I would expect - `256 KB * 1ETH / 64KB = 4 ETH`.

The compressed Fuel Block is - 256KB(all data is new, so compression is bad).
The current implementation will have `3 ETH` debt, assuming the same cost on DA. But if it increases, the debt will be more.

Reply:
`new_cost_per_byte` is the cost per uncompressed bytes.

Absolutely, the compression ratio will be different from block-to-block. We can't get the fully compressed and zipped block, as that doesn't get produced until we actually commit (at which point projecting the cost isn't needed any more).

We could split the difference and get the pre-zipped, compressed block in theory. But I feel like for now, since the blobs include so many blocks, it averages out.

We could look into adding this improvement after we get DA gas costs running and before things get too congested.

Reply:
I think it would be simpler if `new_cost_per_byte` were just the last DA cost per 1 byte(regardless of whether it is compressed or uncompressed). In this case, in the example above, we will charge the user `4 ETH` based on the uncompressed data, assuming the worst-case without compression.